libsolidity/ast/Types.h

```diff
@@ -254,7 +252,10 @@ class Type
 	/// @returns an upper bound on the total storage size required by this type, descending
 	/// into structs and statically-sized arrays. This is mainly to ensure that the storage
 	/// slot allocation algorithm does not overflow, it is not a protection against collisions.
-	virtual bigint storageSizeUpperBound() const { return 1; }
+	bigint storageSizeUpperBound() const { return sizeUpperBound(DataLocation::Storage); }
+	/// @returns an upper bound on the total storage size required by this type
+	/// for the given location
+	virtual bigint sizeUpperBound([[maybe_unused]] DataLocation _loc) const { return 1; }
```

Comment:
As you mentioned below, size is measured in slots or bytes, depending on the data location. It might be confusing, indeed. On the other hand, if it is more natural to count storage in slots and memory in bytes, we might just say that `sizeUpperBound` returns the number of corresponding units. Would you be OK with that?

Reply:
I understand the rationale but I still think it's asking for trouble simply because it's too easy to overlook. I wasn't obvious to me at all when I started looking at this PR. When I see a single function for all three I by default assume that it uses the same units for all of them. I think that just for that reason it might be better to have two different functions - one for storage and one for memory/calldata.

And of course this should be explained in the docstring.

Reply:
I keep it as one single function for now. I am not completely sure how to deal with memory checks vs calldata checks yet.
As soon as we have a complete implementation, I would reconsider splitting up the function into two or three.